File: src/main/java/com/dyonovan/neotech/common/tileentity/generator/TileFluidGenerator.java
package com.dyonovan.neotech.common.tileentity.generator;

import cofh.api.energy.EnergyStorage;
import cofh.api.energy.IEnergyProvider;
import cofh.api.energy.IEnergyReceiver;
import com.dyonovan.neotech.common.blocks.BlockMachine;
import com.dyonovan.neotech.common.blocks.IExpellable;
import com.dyonovan.neotech.common.tileentity.BaseMachine;
import com.dyonovan.neotech.common.tileentity.InventoryTile;
import com.dyonovan.neotech.handlers.BlockHandler;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.server.gui.IUpdatePlayerListBox;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumFacing;
import net.minecraft.world.World;
import net.minecraftforge.fluids.*;

import java.util.ArrayList;
import java.util.List;

import static net.minecraftforge.fluids.FluidContainerRegistry.*;


public class TileFluidGenerator extends BaseMachine implements IUpdatePlayerListBox, IFluidHandler, IEnergyProvider, IExpellable {

    private enum validFuels {
        lava;

        public static boolean contains(String fluid) {
            for (validFuels f : validFuels.values()) {
                if (f.name().equalsIgnoreCase(fluid)) return true;
            }
            return false;
        }
    }

    public FluidTank fluidTank;
    public EnergyStorage energyRF;

    /**
     * Energy Creation per Tick
     */
    private static final int RF_TICK = 80;
    private static final int MB_TICK = 1;
    public static final int TANK_CAPACITY = BUCKET_VOLUME * 10;
    public static final int BUCKET_IN = 0;
    public static final int BUCKET_OUT = 1;

    public TileFluidGenerator() {
        energyRF = new EnergyStorage(10000, RF_TICK);
        inventory = new InventoryTile(2);
        fluidTank = new FluidTank(TANK_CAPACITY);
    }

    @Override
    public void update() {
        if (!this.hasWorldObj()) return;
        World world = this.getWorld();
        if (world.isRemote) return;

        transferEnergy();

        transferFluid();

        if (energyRF.getEnergyStored() + RF_TICK <= energyRF.getMaxEnergyStored() &&
                fluidTank.getFluid() != null && fluidTank.getFluidAmount() >= MB_TICK) {
            energyRF.modifyEnergyStored(RF_TICK);
            fluidTank.drain(MB_TICK, true);
            BlockMachine.setState(worldObj, pos, BlockHandler.fluidGeneratorActive);
            world.markBlockForUpdate(this.pos);
        } else if(worldObj.getBlockState(pos).getBlock() == BlockHandler.fluidGeneratorActive)
            BlockMachine.setState(worldObj, pos, BlockHandler.fluidGenerator);

    }

    private void transferEnergy() {
        List<EnumFacing> availDir = new ArrayList<>();
        if (energyRF.getEnergyStored() > 0) {
            for (EnumFacing dir : EnumFacing.VALUES) {
                TileEntity tile = getWorld().getTileEntity(this.pos.offset(dir));
                if (tile instanceof IEnergyReceiver) availDir.add(dir);
            }
        }
        if (availDir.size() <= 0) return;
        int availRF = Math.min(energyRF.getEnergyStored() / availDir.size() , energyRF.getMaxExtract() / availDir.size());
        for (EnumFacing dir : availDir) {
            TileEntity tile = getWorld().getTileEntity(this.pos.offset(dir));
            energyRF.extractEnergy(((IEnergyReceiver) tile).receiveEnergy(dir.getOpposite(), energyRF.extractEnergy(availRF, true), false), false);
        }
        getWorld().markBlockForUpdate(this.pos);
    }

    private void transferFluid() {
        if (inventory.getStackInSlot(BUCKET_IN) == null ||
                !isFilledContainer(inventory.getStackInSlot(BUCKET_IN))) return;

        FluidStack fluidIn = getFluidForFilledItem(inventory.getStackInSlot(BUCKET_IN));
        if (fluidIn == null) return;

        if (!(validFuels.contains(fluidIn.getFluid().getName()))) return;
        if (fluidTank.getFluid() != null && fluidTank.getFluid().getFluid() != fluidIn.getFluid()) return;
        if (fluidTank.getFluid() != null && fluidTank.getFluid().amount + fluidIn.amount > fluidTank.getCapacity()) return;


        fluidTank.fill(fluidIn, true);

        if (inventory.getStackInSlot(BUCKET_OUT) == null)
            inventory.setStackInSlot(drainFluidContainer(inventory.getStackInSlot(BUCKET_IN)),BUCKET_OUT);
        else inventory.modifyStack(BUCKET_OUT, 1);
        inventory.modifyStack(BUCKET_IN, -1);


        this.getWorld().markBlockForUpdate(this.pos);
    }



    /*******************************************************************************************************************
     *************************************** Fluid Functions ***********************************************************
     *******************************************************************************************************************/

    @Override
    public int fill(EnumFacing from, FluidStack resource, boolean doFill) {
            int actual =  fluidTank.fill(resource, doFill);
            this.getWorld().markBlockForUpdate(this.pos);
            return actual;
    }

    @Override
    public FluidStack drain(EnumFacing from, FluidStack resource, boolean doDrain) {
        if (resource == null || !resource.isFluidEqual(fluidTank.getFluid()))
        {
            return null;
        }
        FluidStack actual = fluidTank.drain(resource.amount, doDrain);
        this.getWorld().markBlockForUpdate(this.pos);
        return actual;
    }

    @Override
    public FluidStack drain(EnumFacing from, int maxDrain, boolean doDrain) {
        FluidStack actual = fluidTank.drain(maxDrain, doDrain);
        this.getWorld().markBlockForUpdate(this.pos);
        return actual;
    }

    @Override
    public boolean canFill(EnumFacing from, Fluid fluid) {
        return true;
    }

    @Override
    public boolean canDrain(EnumFacing from, Fluid fluid) {
        return false;
    }

    @Override
    public FluidTankInfo[] getTankInfo(EnumFacing from) {
        return new FluidTankInfo[] { fluidTank.getInfo() };
    }

    /*******************************************************************************************************************
     ************************************** Energy Functions ***********************************************************
     *******************************************************************************************************************/

    @Override
    public int extractEnergy(EnumFacing from, int maxExtract, boolean simulate) {
        int actual = energyRF.extractEnergy(maxExtract, simulate);
        if (actual > 0) this.getWorld().markBlockForUpdate(this.pos);
        return actual;
    }

    @Override
    public int getEnergyStored(EnumFacing from) {
        return energyRF.getEnergyStored();
    }

    @Override
    public int getMaxEnergyStored(EnumFacing from) {
        return energyRF.getMaxEnergyStored();
    }

    @Override
    public boolean canConnectEnergy(EnumFacing from) {
        return true;
    }

    @Override
    public void spawnActiveParticles(double x, double y, double z) {
    }

    /*******************************************************************************************************************
     ************************************** Inventory Functions ********************************************************
     *******************************************************************************************************************/

    @Override
    public int[] getSlotsForFace(EnumFacing side) {
        return new int[] {0,1};
    }

    @Override
    public boolean canInsertItem(int index, ItemStack itemStackIn, EnumFacing direction) {
        return index == BUCKET_IN;
    }

    @Override
    public boolean canExtractItem(int index, ItemStack stack, EnumFacing direction) {
        return index == BUCKET_OUT;
    }

    @Override
    public int getSizeInventory() {
        return 2;
    }

    @Override
    public boolean isItemValidForSlot(int index, ItemStack stack) {
        return isFilledContainer(inventory.getStackInSlot(BUCKET_IN));
    }

    @Override
    public int getField(int id) {
        return 0;
    }

    @Override
    public void setField(int id, int value) {
    }

    @Override
    public int getFieldCount() {
        return 0;
    }

    @Override
    public void clear() {
        for (int i = 0; i < inventory.getSizeInventory(); i++) {
            inventory.setStackInSlot(null, i);
        }
    }

    /*******************************************************************************************************************
     **************************************** Tile Functions ***********************************************************
     *******************************************************************************************************************/

    @Override
    public void readFromNBT(NBTTagCompound tag) {
        super.readFromNBT(tag);
        energyRF.readFromNBT(tag);
        inventory.readFromNBT(tag, this);
        fluidTank.readFromNBT(tag);
    }

    @Override
    public void writeToNBT(NBTTagCompound tag) {
        super.writeToNBT(tag);
        energyRF.writeToNBT(tag);
        inventory.writeToNBT(tag);
        fluidTank.writeToNBT(tag);
    }
}

File: src/main/scala/com/dyonovan/neotech/common/tiles/machines/TileFluidGenerator.scala
package com.dyonovan.neotech.common.tiles.machines

import cofh.api.energy.{IEnergyReceiver, EnergyStorage}
import com.dyonovan.neotech.common.tiles.AbstractMachine
import com.dyonovan.neotech.registries.FluidFuelValues
import com.teambr.bookshelf.common.blocks.properties.PropertyRotation
import com.teambr.bookshelf.util.InventoryUtils
import net.minecraft.inventory.{IInventory, Container}
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing
import net.minecraftforge.fluids._

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Dyonovan
  * @since August 21, 2015
  */
class TileFluidGenerator extends AbstractMachine with IFluidHandler {

    final val RF_TICK = 80
    energy = new EnergyStorage(100000)
    val tank = new FluidTank(FluidContainerRegistry.BUCKET_VOLUME * 10)

    override def doWork(): Unit = {

        //Generate
        if (this.values.burnTime > 0) {
            this.values.burnTime -= 1
            energy.receiveEnergy(RF_TICK, false)
            worldObj.markBlockForUpdate(pos)
        } else if (energy.getEnergyStored < energy.getMaxEnergyStored && tank.getFluid != null && tank.getFluidAmount >= 100) {
            this.values.burnTime = (FluidFuelValues.getFluidFuelValue(tank.getFluid.getFluid.getName) * .1).toInt
            this.values.currentItemBurnTime = this.values.burnTime
            tank.drain(100, true)
        }

        //Transfer
        if (energy.getEnergyStored > 0) {
            for (i <- EnumFacing.values()) {
                worldObj.getTileEntity(pos.offset(i)) match {
                    case tile: IEnergyReceiver =>
                        val want = tile.receiveEnergy(i.getOpposite, energy.getEnergyStored, true)
                        if (want > 0) {
                            val actual = energy.extractEnergy(want, false)
                            tile.receiveEnergy(i.getOpposite, actual, false)
                            worldObj.markBlockForUpdate(pos)
                        }
                    case _ =>
                }
            }
        }
    }


    override def spawnActiveParticles(x: Double, y: Double, z: Double): Unit = {}

    override def tryInput() : Unit = {
        for(dir <- EnumFacing.values) {
            if(canInputFromSide(dir, worldObj.getBlockState(pos).getValue(PropertyRotation.FOUR_WAY))) {
                worldObj.getTileEntity(pos.offset(dir)) match {
                    case otherTank : IFluidHandler =>
                        if(otherTank.getTankInfo(dir.getOpposite)(0).fluid != null && canFill(dir, otherTank.getTankInfo(dir.getOpposite)(0).fluid.getFluid))
                            fill(dir, otherTank.drain(dir.getOpposite, 1000, true), doFill = true)
                    case _ =>
                }
            }
        }
    }

    /**
      * Get the output of the recipe
      *
      * @param stack The input
      * @return The output
      */
    override def recipe(stack: ItemStack): ItemStack = null

    /**
      * Used to output the redstone single from this structure
      *
      * Use a range from 0 - 16.
      *
      * 0 Usually means that there is nothing in the tile, so take that for lowest level. Like the generator has no energy while
      * 16 is usually the flip side of that. Output 16 when it is totally full and not less
      *
      * @return int range 0 - 16
      */
    override def getRedstoneOutput: Int = InventoryUtils.calcRedstoneFromInventory(this)

    override def drain(from: EnumFacing, resource: FluidStack, doDrain: Boolean): FluidStack = drain(from, resource, doDrain)

    override def drain(from: EnumFacing, maxDrain: Int, doDrain: Boolean): FluidStack = {
        val fluidAmount = tank.drain(maxDrain, false)
        if (fluidAmount != null && doDrain)
            tank.drain(maxDrain, true)
        worldObj.markBlockForUpdate(pos)

        fluidAmount
    }

    override def canFill(from: EnumFacing, fluid: Fluid): Boolean = {
        (tank.getFluid == null || tank.getFluid.getFluid == fluid) && FluidFuelValues.isFluidFuel(fluid.getName)
    }


    override def canDrain(from: EnumFacing, fluid: Fluid): Boolean = false

    override def fill(from: EnumFacing, resource: FluidStack, doFill: Boolean): Int = {
        if (canFill(from, resource.getFluid)) {
            if (tank.fill(resource, false) > 0) {
                val actual = tank.fill(resource, doFill)
                worldObj.markBlockForUpdate(pos)
                return actual
            }
        }
        0
    }

    override def getTankInfo(from: EnumFacing): Array[FluidTankInfo] = Array(tank.getInfo)

    override def receiveEnergy(from: EnumFacing, maxReceive: Int, simulate: Boolean): Int = 0

    override def extractEnergy(from: EnumFacing, maxExtract: Int, simulate: Boolean): Int = {
        val actual = energy.extractEnergy(maxExtract, simulate)
        worldObj.markBlockForUpdate(pos)
        actual
    }

    override def writeToNBT(tag: NBTTagCompound): Unit = {
        super.writeToNBT(tag)
        tank.writeToNBT(tag)
    }

    override def readFromNBT(tag: NBTTagCompound): Unit = {
        super.readFromNBT(tag)
        tank.readFromNBT(tag)
    }
}
